rules/firestore.json

```diff
@@ -92,39 +92,102 @@ service cloud.firestore {
-                             partHasStarted(resource.data) &&
-                             userIsPlayer(resource.data) &&
-                             preservesField('typeGame') &&
-                             preservesField('playerZero') &&
-                             preservesField('playerOne') &&
-                             preservesField('beginning') &&
-                             (preservesField('turn') || safeTurnUpdate(resource.data.turn, request.resource.data.turn)) &&
-                             (safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate));
+            allow update: if chosenOpponentStartsPart(id);
+            // A player can perform a move
+            allow update: if isValidMovePartUpdate();
+            // A player can do other requests such as take backs, etc. if there is no move done
+            allow update: if isValidPartUpdateWithoutMove();
             // All verified user can read parts (needed for the lobby)
             allow read: if userIsVerified();
         }
-        function userIsChosenOpponentInJoiner(partId) {
-            let joiner = get(/databases/$(database)/documents/joiners/$(partId)).data;
-            return hasField(joiner, 'chosenOpponent') && joiner.chosenOpponent.id == request.auth.uid;
+        function isValidMovePartUpdate() {
+            let decreasesPlayerZeroTime = debug(safeTimeDecrease(resource.data.playerZero, 'remainingMsForZero'));
+            let preservesPlayerZeroTime = debug(preservesField('remainingMsForZero'));
+            let decreasesPlayerOneTime = debug(safeTimeDecrease(resource.data.playerOne, 'remainingMsForOne'));
+            let preservesPlayerOneTime = debug(preservesField('remainingMsForOne'));
+            let safeTimeDecrease = (resource.data.turn == 0 && preservesPlayerOneTime && preservesPlayerZeroTime) ||
+                (decreasesPlayerZeroTime && preservesPlayerOneTime) ||
+                (preservesPlayerZeroTime && decreasesPlayerOneTime);
+            let isUserTurn = (resource.data.turn % 2 == 0 && userIsPlayerZero(resource.data)) ||
+                (resource.data.turn % 2 == 1 && userIsPlayerOne(resource.data));
+            return isValidPartUpdateDuringGame() &&
+                safeTimeDecrease &&
+                safeTurnIncrease(resource.data.turn, request.resource.data.turn) &&
+                safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate) &&
+                isUserTurn;
+        }
+        function isValidPartUpdateWithoutMove() {
+            return isValidPartUpdateDuringGame() &&
+                safeTimeUpdate(resource.data.playerZero, 'remainingMsForZero') &&
+                safeTimeUpdate(resource.data.playerOne, 'remainingMsForOne') &&
+                (preservesField('turn') || safeTurnDecrease(resource.data.turn, request.resource.data.turn)) &&
+                safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate);
+        }
```

Comment:
les update non move, c'est bien des requêtes ou leurs réponses, right (c'qui indique qu'y a le chams request dans les deux) ? J'imagien qu'il faut pas d'excès mais là faudrait par exemple au moins vérifier que j'ignore pas ta requête en l'écrasant avec la mienne. Ou qu'une demande doit être suivie de sa réponse ? (ça se tickette, si besoin, j'y serais favorable :D)

Reply:
tiens en fait, dans le genre, on a un check qui dit "you cannot just say OKAY PART IS A DRAW" (vu que contrairements aux victoires non checkable côté sécu, celles là sont des réponses à une demande, donc pas forçable unilatéralement ?

Reply:
TODO